src/main/java/org/truffleruby/core/rope/RopeNodes.java

```diff
@@ -251,10 +238,145 @@ private Rope makeSubstringNon7Bit(Encoding encoding, Rope base, int byteOffset,
             }
         }
 
-        protected static boolean sameAsBase(Rope base, int byteLength) {
-            // A SubstringRope's byte length is not allowed to be larger than its child. Thus, if it has the same
-            // byte length as its child, it must be logically equivalent to the child.
-            return byteLength == base.byteLength();
+        protected static boolean is7Bit(CodeRange codeRange) {
+            return codeRange == CR_7BIT;
+        }
+
+    }
+
+    @NodeChildren({
+            @NodeChild(type = RubyNode.class, value = "encoding"),
+            @NodeChild(type = RubyNode.class, value = "bytes")
+    })
+    @ImportStatic(RopeGuards.class)
+    public abstract static class CalculateAttributesNode extends RubyNode {
+
+        public static CalculateAttributesNode create() {
+            return RopeNodesFactory.CalculateAttributesNodeGen.create(null, null);
+        }
+
+        abstract StringAttributes executeCalculateAttributes(Encoding encoding, byte[] bytes);
+
+        @Specialization(guards = "isEmpty(bytes)")
+        StringAttributes calculateAttributesEmpty(Encoding encoding, byte[] bytes,
```

Comment:
It's unusual to see so many package-private methods - what's going on here?

Reply:
I think it's less noisy like that and doesn't really lose anything.
`public` is harmful (`@Specialization` should never be used outside the Node).